### ISSUE TITLE
PHP 8 is_callable changes

### ISSUE THREAD

**Issue by Spuds** [elkarte/Elkarte#3419]
I started to test 1.1.7 under PHP 8 to see what issues we may have, and if they are not to extensive, make the updates.  That way why we get ahead of users / hosts with CUD (Compulsive Upgrade Disorder)  since 8 is now released.

So the first issue is our buddy ```is_callable(array($class_string, $method_string))```  That will not longer work as expected:

> The ability to call non-static methods statically has been removed. Thus is_callable() will fail when checking for a non-static method with a classname (must check with an object instance).

You need a try/catch with ```$bla = new $class_string()``` and then ```is_callable(array($bla, $method_string))``` Which I don't like having to instantiate a class to check and then destroy the class when done.

Other options I've tried

```if (class_exists($class_string) && method_exists($method_string))``` This works but with the shortcoming of a method existing does not mean you can call it directly.

```if (class_exists($class_string) && in_array($method_string, get_class_methods($class_string)))``` From my testing as long as you are doing ```get_class_methods``` on the class name, it only returns public ones (including inherited), so this behavior is closer to the is_callable behavior.  

other ideas? I think the last of these is the cleanest so far but ?

**Comment by emanuele45** [comment]
hmm... you assign it to me to make me feel responsible? Sneaky bastard! :P

Unless I missed something (and I'm pretty sure I did), this situation arises only in the dispatcher?
Well, for the dispatcher, instantiating the controller before knowing if the action is available is a bit meh, but the case in which the controller instantiated would have to be destroyed and re-instantiated should be pretty unusual, the only case should technically be an addon with a broken controller that defaults to the "home page" (i.e. board index).
In that case, I'd be almost tempted to say screw checks: if the controller exists it has to have a callable method.

Is there any other place we have the same issue?

**Comment by Spuds** [comment]
LOL I figured that would get your attention, what the hell are you doing Spuds.

Its the dispatcher, the hook processor and the special front page hook areas.   All of the other uses that I've seen are using the call on an object, not a string.  I changed the other areas to use the ```get_class_methods()``` call and its working as expected.  But to your point in 2.0 I think we could rework that and gain some efficiency.   Its pretty protective now and that actually made finding the problem difficult since it did not toss or log any error.

The only other 8.0 issue of note is in the event manager.  PHP 8 will stop dead if you pass parameters to a method that accepts none (and presumably wrongly named / ordered ones).   So I did a little dance with
```
				// Lets check on the required method parameters
				try
				{
					$r = new ReflectionMethod($class_name, $method_name);
					$number_params = $r->getNumberOfParameters();
					unset($r);
				}
				catch (\Exception $e)
				{
					$number_params = 0;
				}

                               if (empty($dependencies) || empty($number_params))
                                     bla bla bla
```
Which will ensure no parameters are sent to a method that uses none.   Dirty but seemed like an easy fix.  In 2.0 could use this to ensure the parameter names match what is expected, log errors, etc etc

Anyway have to run to the store now, my neighbor has the old corona-mona and needs some stuff so off I go!
